packages/connect-web/src/impl/core-in-suite-desktop.ts

```diff
@@ -0,0 +1,172 @@
+    ConnectSettingsWeb,
+    Manifest,
+    Response,
+} from '@trezor/connect/src/types';
+import { ConnectFactoryDependencies, factory } from '@trezor/connect/src/factory';
+
+import { parseConnectSettings } from '../connectSettings';
+import { Login } from '@trezor/connect/src/types/api/requestLogin';
+
+/**
+ * Base class for CoreInPopup methods for TrezorConnect factory.
+ * This implementation is directly used here in connect-web, but it is also extended in connect-webextension.
+ */
+export class CoreInSuiteDesktop implements ConnectFactoryDependencies<ConnectSettingsWeb> {
+    public eventEmitter = new EventEmitter();
+    protected _settings: ConnectSettings;
+    private ws?: WebSocket;
+
+    public constructor() {
+        this._settings = parseConnectSettings();
+    }
+
+    public manifest(data: Manifest) {
+        this._settings = parseConnectSettings({
+            ...this._settings,
+            manifest: data,
+        });
+    }
+
+    public dispose() {
+        this.eventEmitter.removeAllListeners();
+        this._settings = parseConnectSettings();
+
+        return Promise.resolve(undefined);
+    }
+
+    public cancel(_error?: string) {}
+
+    private async handshake() {
+        return Promise.race([
+            new Promise<void>(resolve => {
+                const listener = (event: WebSocketEventMap['message']) => {
+                    if (event.data === 'handshake') {
+                        resolve();
+                        removeEventListener('message', listener);
+                    }
+                };
+                this.ws?.addEventListener('message', listener);
+                this.ws?.send('handshake');
+            }),
+            new Promise<void>((_resolve, reject) => {
+                setTimeout(() => {
+                    reject(new Error('Handshake timeout'));
+                }, 1000);
+            }),
+        ]);
+    }
+    public async init(settings: Partial<ConnectSettingsPublic> = {}): Promise<void> {
+        const newSettings = parseConnectSettings({
+            ...this._settings,
+            ...settings,
+        });
+
+        // defaults
+        if (!newSettings.transports?.length) {
+            newSettings.transports = ['BridgeTransport', 'WebUsbTransport'];
+        }
+        this._settings = newSettings;
+
+        this.ws = new WebSocket('ws://localhost:8090');
+        this.ws.addEventListener('error', console.error);
+
+        await new Promise(resolve => {
+            this.ws?.addEventListener('opened', resolve);
+        });
+        return this.handshake();
+    }
+
+    /**
+     * 1. opens popup
+     * 2. sends request to popup where the request is handled by core
+     * 3. returns response
+     */
+    public async call(params: CallMethodPayload): Promise<CallMethodAnyResponse> {
+        try {
+            await this.handshake();
+
+            this.ws?.send(
+                JSON.stringify({
+                    type: IFRAME.CALL,
```

Comment:
yes. I think we already have couple of places where we use the `send and listen for event` communication mechanism. It would be nice to find some common abstraction for it